experimental/collator/src/lib.rs

```diff
@@ -0,0 +1,4649 @@
+    }
+}
+
+impl Default for CollationElement32 {
+    fn default() -> Self {
+        CollationElement32(1) // NO_CE32
+    }
+}
+
+/// A collation element is a 64-bit value.
+///
+/// The high 32 bits are the primary weight.
+/// The high 16 bits of the low 32 bits are the secondary weight.
+/// The high 2 bits of the low 16 bits are the case bits.
+/// The high 2 bits of the low 8 bit are the quaternary weight.
+/// The low 6 bits of the second to lowest 8 bits and the low 6
+/// bits for the (bitwise discontiguous) tertiary weight.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) struct CollationElement(u64);
+
+impl CollationElement {
+    #[inline(always)]
+    pub fn new(bits: u64) -> Self {
+        CollationElement(bits)
+    }
+
+    #[inline(always)]
+    pub fn new_from_ule(ule: RawBytesULE<8>) -> Self {
+        CollationElement(u64::from_unaligned(ule))
+    }
+
+    #[inline(always)]
+    pub fn new_from_primary(primary: u32) -> Self {
+        CollationElement((u64::from(primary) << 32) | COMMON_SEC_AND_TER_CE)
+    }
+
+    #[inline(always)]
+    pub fn new_implicit_from_char(c: char) -> Self {
+        // Collation::unassignedPrimaryFromCodePoint
+        // Create a gap before U+0000. Use c-1 for [first unassigned].
+        let mut c_with_offset = u32::from(c) + 1;
+        // Fourth byte: 18 values, every 14th byte value (gap of 13).
+        let mut primary: u32 = 2 + (c_with_offset % 18) * 14;
+        c_with_offset /= 18;
+        // Third byte: 254 values
+        primary |= (2 + (c_with_offset % 254)) << 8;
+        c_with_offset /= 254;
+        // Second byte: 251 values 04..FE excluding the primary compression bytes.
+        primary |= (4 + (c_with_offset % 251)) << 16;
+        // One lead byte covers all code points (c < 0x1182B4 = 1*251*254*18).
+        primary |= u32::from(UNASSIGNED_IMPLICIT_BYTE) << 24;
+        CollationElement::new_from_primary(primary)
+    }
+
+    #[inline(always)]
+    pub fn clone_with_non_primary_zeroed(&self) -> Self {
+        CollationElement(self.0 & 0xFFFFFFFF00000000)
+    }
+
+    /// Get the primary weight
+    #[inline(always)]
+    pub fn primary(&self) -> u32 {
+        (self.0 >> 32) as u32
+    }
+
+    /// Get the non-primary weight
+    #[inline(always)]
+    pub fn non_primary(&self) -> NonPrimary {
+        NonPrimary::new(self.0 as u32)
+    }
+
+    /// Get the secondary weight
+    #[inline(always)]
+    pub fn secondary(&self) -> u16 {
+        self.non_primary().secondary()
+    }
+    #[inline(always)]
+    pub fn quaternary(&self) -> u32 {
+        self.non_primary().quaternary()
+    }
+    #[inline(always)]
+    pub fn tertiary_ignorable(&self) -> bool {
+        self.non_primary().tertiary_ignorable()
+    }
+    #[inline(always)]
+    pub fn either_half_zero(&self) -> bool {
+        self.primary() == 0 || (self.0 as u32) == 0
+    }
+
+    #[inline(always)]
+    pub const fn const_default() -> CollationElement {
+        CollationElement(0x101000100)
```

Comment:
suggestion: refactor code for more clarity

This magic number seems important, based on usages here and below.  Maybe give the constant a name, a quick Rust doc string, and IIUC, also preferably define it on the right-hand side of the `const` definition in terms of bit shifts and the relevant other consts `NO_CE_SECONDARY` and `NO_CE_TERTIARY`.  It looks like you have a primary weight of 1, and that would become more self-evident.

Reply:
Done.